File: app/src/main/java/com/xymonitor/app/Inspector.kt
package com.xymonitor.app

data class InspectOutcome(
    val ok: Boolean,
    val baseline: Boolean,
    val changed: Boolean,
    val firstId: String,
    val error: String?,
)

object Inspector {
    fun compare(previousFirstId: String, currentFirstId: String): InspectOutcome {
        if (currentFirstId.isBlank()) {
            return fail("第一页没有商品")
        }
        if (previousFirstId.isBlank()) {
            return InspectOutcome(
                ok = true,
                baseline = true,
                changed = false,
                firstId = currentFirstId,
                error = null,
            )
        }
        val changed = previousFirstId != currentFirstId
        return InspectOutcome(
            ok = true,
            baseline = false,
            changed = changed,
            firstId = currentFirstId,
            error = null,
        )
    }

    fun fail(message: String): InspectOutcome {
        return InspectOutcome(
            ok = false,
            baseline = false,
            changed = false,
            firstId = "",
            error = message,
        )
    }
}

File: server/inspector.py
from dataclasses import dataclass


@dataclass(frozen=True)
class InspectOutcome:
    ok: bool
    baseline: bool
    changed: bool
    first_id: str
    error: str | None


def compare(previous_first_id: str, current_first_id: str) -> InspectOutcome:
    if not current_first_id:
        return fail("第一页没有商品")
    if not previous_first_id:
        return InspectOutcome(
            ok=True,
            baseline=True,
            changed=False,
            first_id=current_first_id,
            error=None,
        )
    return InspectOutcome(
        ok=True,
        baseline=False,
        changed=previous_first_id != current_first_id,
        first_id=current_first_id,
        error=None,
    )


def fail(message: str) -> InspectOutcome:
    return InspectOutcome(
        ok=False,
        baseline=False,
        changed=False,
        first_id="",
        error=message,
    )
